container/memory_store.go

```diff
@@ -73,19 +84,60 @@ func (c *memoryStore) First(filter StoreFilter) *Container {
 // ApplyAll calls the reducer function with every container in the store.
 // This operation is asyncronous in the memory store.
 func (c *memoryStore) ApplyAll(apply StoreReducer) {
-	c.Lock()
-	defer c.Unlock()
+	c.globalLock.Lock()
+	defer c.globalLock.Unlock()
 
 	wg := new(sync.WaitGroup)
 	for _, cont := range c.s {
```

Comment:
I would make sure you only hold the global lock while reading the map.
No need to wait for the waitgroup to unlock, which may even cause a deadlock if for some reason someone's reducder is calling back into the store.